core/instance.go

```diff
@@ -18,53 +44,255 @@ type instanceTypeInformation struct {
+				candidate.pricing.spot[availabilityZone], "<=",
+				i.price)
+		} else {
+			logger.Println("price too high, skipping", candidate.instanceType)
+			continue
+		}
+
+		if candidate.vCPU >= existing.vCPU {
+			logger.Println("CPU compatible, continuing evaluation")
+		} else {
+			logger.Println("Insuficient CPU cores, skipping", candidate.instanceType)
+			continue
+		}
+
+		if candidate.memory >= existing.memory {
+			logger.Println("memory compatible, continuing evaluation")
+		} else {
+			logger.Println("memory incompatible, skipping", candidate.instanceType)
+			continue
+		}
+
+		// Here we check the storage compatibility, with the following evaluation
+		// criteria:
+		// - speed: don't accept spinning disks when we used to have SSDs
+		// - number of volumes: the new instance should have enough volumes to be
+		//   able to attach all the instance store device mappings defined on the
+		//   original instance
+		// - volume size: each of the volumes should be at least as big as the
+		//   original instance's volumes
+
+		if attachedVolumesNumber > 0 {
+			logger.Println("Checking the new instance's ephemeral storage",
+				"configuration because the initial instance had attached",
+				"ephemeral instance store volumes")
+
+			if candidate.instanceStoreDeviceCount >= attachedVolumesNumber {
+				logger.Println("instance store volume count compatible,",
+					"continuing evaluation")
+			} else {
+				logger.Println("instance store volume count incompatible, skipping",
+					candidate.instanceType)
+				continue
+			}
+
+			if candidate.instanceStoreDeviceSize >= existing.instanceStoreDeviceSize {
+				logger.Println("instance store volume size compatible,",
+					"continuing evaluation")
+			} else {
+				logger.Println("instance store volume size incompatible, skipping",
+					candidate.instanceType)
+				continue
+			}
+
+			// Don't accept ephemeral spinning disks if the original instance has
+			// ephemeral SSDs, but accept spinning disks if we had those before.
+			if candidate.instanceStoreIsSSD ||
+				(candidate.instanceStoreIsSSD == existing.instanceStoreIsSSD) {
+				logger.Println("instance store type(SSD/spinning) compatible,",
+					"continuing evaluation")
+			} else {
+				logger.Println("instance store type(SSD/spinning) incompatible,",
+					"skipping", candidate.instanceType)
+				continue
+			}
+		}
+
+		if compatibleVirtualization(*i.VirtualizationType,
+			candidate.virtualizationTypes) {
+			logger.Println("virtualization compatible, continuing evaluation")
+		} else {
+			logger.Println("virtualization incompatible, skipping",
+				candidate.instanceType)
+			continue
+		}
+
+		// checking how many spot instances of this type we already have, so that
+		// we can see how risky it is to launch a new one.
+		spotInstanceCount := i.asg.alreadyRunningSpotInstanceCount(
+			candidate.instanceType, availabilityZone)
+
+		// We skip it in case we have more than 20% instances of this type already
+		// running
```

Comment:
I'm also fairly new to golang, so no worries.

Why is it a problem in tests?
For the log tool, I don't have any preference I have heard of https://github.com/Sirupsen/logrus and https://github.com/golang/glog, but apparently there are many others such as: https://github.com/inconshreveable/log15 or https://github.com/op/go-logging.

Considering the fairly basic usage (having log levels), I think any could do, it's only a matter of preference and/or evolution afterwards.